File: ruby/src/data-structures/priority_queue.rb
# frozen_string_literal: true

EMPTY_PRIORITY_QUEUE_EXCEPTION = 'Priority queue is empty'

module DataStructures
  # Priority goes high to low.
  # The predicate lambda should return 1, 0, or -1.
  # If predicate.call(a, b) = 1, then a has a higher priority than b.

  class ArrayPriorityQueue
    def initialize(predicate)
      @elems = []
      @predicate = predicate
    end

    def peek
      return raise EMPTY_PRIORITY_QUEUE_EXCEPTION if @elems.length.zero?

      @elems[0]
    end

    def enqueue(value)
      @elems.push(value)
      @elems.sort! { |a, b| @predicate.call(a, b) }.reverse!
    end

    def dequeue
      return raise EMPTY_PRIORITY_QUEUE_EXCEPTION if @elems.length.zero?

      @elems.shift
    end

    def size
      @elems.length
    end

    def clear
      @elems = []
    end

    def empty?
      @elems.length.zero?
    end
  end

  class LinkedPriorityQueue
    attr_accessor :size

    class QueueNode
      attr_accessor :value, :next

      def initialize(value, next_node)
        @value = value
        @next = next_node
      end
    end

    def initialize(predicate)
      @predicate = predicate
      @head = nil
      @size = 0
    end

    def peek
      return raise EMPTY_PRIORITY_QUEUE_EXCEPTION if @head.nil?

      @head.value
    end

    def enqueue(value)
      node = QueueNode.new(value, nil)

      if @head.nil?
        @head = node
        @size += 1
        return
      end

      if @size == 1
        if @predicate.call(value, @head.value) >= 0
          node.next = @head
          @head = node
        else
          @head.next = node
        end
        @size += 1
        return
      end

      if @predicate.call(value, @head.value) >= 0
        node.next = @head
        @head = node
        @size += 1
        return
      end

      previous = @head
      current = @head.next
      until current.nil?
        break if @predicate.call(value, current.value) >= 0

        previous = current
        current = current.next
      end
      node.next = current
      previous.next = node
      @size += 1
    end

    def dequeue
      return raise EMPTY_PRIORITY_QUEUE_EXCEPTION if @head.nil?

      value = @head.value
      @head = @head.next
      @size -= 1

      value
    end

    def clear
      @head = nil
      @size = 0
    end

    def empty?
      @head.nil?
    end
  end
end

File: typescript/src/data-structures/priority-queue.ts
import { Queue, QueueNode } from "./queue";

export abstract class PriorityQueue<T> extends Queue<T> {
  /**
   * Priority goes high to low.
   * If predicate(a, b) = 1, then a has a higher priority than b.
   */
  protected predicate: (a: T, b: T) => 1 | 0 | -1;

  constructor(predicate: (a: T, b: T) => 1 | 0 | -1) {
    super();
    this.predicate = predicate;
  }
}

export class ArrayPriorityQueue<T> extends PriorityQueue<T> {
  private elems: T[];

  constructor(predicate: (a: T, b: T) => 1 | 0 | -1) {
    super(predicate);
    this.elems = [];
  }

  peek() {
    if (this.elems.length === 0) {
      throw new Error("Queue is empty");
    }
    return this.elems[0];
  }

  enqueue(elem: T) {
    this.elems.push(elem);
    this.elems = this.elems.sort(this.predicate).reverse();
  }

  dequeue() {
    if (this.elems.length === 0) {
      throw new Error("Queue is empty");
    }
    return this.elems.shift();
  }

  size() {
    return this.elems.length;
  }

  clear() {
    this.elems = [];
  }
}

export class LinkedPriorityQueue<T> extends PriorityQueue<T> {
  private head: QueueNode<T>;
  private _size: number;

  constructor(predicate: (a: T, b: T) => 1 | 0 | -1) {
    super(predicate);
    this.head = undefined;
    this._size = 0;
  }

  peek() {
    if (!this.head) {
      throw new Error("Queue is empty");
    }
    return this.head.value;
  }

  enqueue(value: T) {
    const node: QueueNode<T> = {
      value,
      next: undefined
    };

    if (!this.head) {
      this.head = node;
      this._size++;
      return;
    }

    if (this.size() === 1) {
      if (this.predicate(value, this.head.value) >= 0) {
        node.next = this.head;
        this.head = node;
      } else {
        this.head.next = node;
      }
      this._size++;
      return;
    }

    if (this.predicate(value, this.head.value) >= 0) {
      node.next = this.head;
      this.head = node;
      this._size++;
      return;
    }

    let previousNode = this.head;
    let currentNode = this.head.next;
    while (currentNode) {
      if (this.predicate(value, currentNode.value) >= 0) {
        break;
      }
      previousNode = currentNode;
      currentNode = currentNode.next;
    }
    node.next = currentNode;
    previousNode.next = node;
    this._size++;
  }

  dequeue() {
    if (!this.head) {
      throw new Error("Queue is empty");
    }

    const { value } = this.head;
    this.head = this.head.next;
    this._size--;

    return value;
  }

  size() {
    return this._size;
  }

  clear() {
    this.head = undefined;
    this._size = 0;
  }
}
